### PR TITLE
[9.0] Migrate web_m2x_options to 9.0

### PR DESCRIPTION
**Pull request by zakiuu**

### CODE REVIEW

**Comment by sbidoul** [comment]
@zakiuu have you handled this https://github.com/OCA/web/pull/416#discussion_r78050624 from the other PR?

File: web_m2x_options/static/src/js/form.js
Comment (lmignon) [inline web_m2x_options/static/src/js/form.js]: At top of your file add 

``` javascript
var $ = require("$");
```

And replace `JQuery.when` with `$.when`

File: web_m2x_options/static/src/js/form.js
Comment (lmignon) [inline web_m2x_options/static/src/js/form.js]: you should call a method declared on `FieldMany2ManyTags` so it will be possible to override it.  (??? `on_badge_click` ???)

File: web_m2x_options/static/src/js/form.js
Comment (lmignon) [inline web_m2x_options/static/src/js/form.js]: s/this/e.target

File: web_m2x_options/static/src/js/form.js
Comment (lmignon) [inline web_m2x_options/static/src/js/form.js:32]: PlZ add a `var _ = require('_');` since underscore is used in this JS file

**Review by lmignon — CHANGES_REQUESTED** [review]
I don't know what's going wrong but the module seems to not work
image

File: web_m2x_options/static/src/js/form.js
Comment (sbidoul) [inline web_m2x_options/static/src/js/form.js]: upstream, it's 160 not 80.

BTW, the code is so different from the upstream `get_search_result` that I wonder if it has diverged elsewhere.

**Review by sbidoul — COMMENTED** [review]
@zakiuu don't forget this one, not a big deal but let's have the same behaviour as upstream.

**Comment by lmignon** [comment]
@zakiuu It seems that you have to report a bug to Odoo. Your implementation can only work with Odoo Enterprise. Indeed in the community version of `web` the  widget `FieldMany2One` that you extend at https://github.com/OCA/web/pull/418/files#diff-d93aa186928a77371c865d1748af9b38R55 is not exposed by the js module `web.form_relational`  https://github.com/odoo/odoo/blob/9.0/addons/web/static/src/js/views/form_relational_widgets.js#L1734

The return should be 

``` javascript
return {
    FieldMany2ManyTags: FieldMany2ManyTags,
    AbstractManyField: AbstractManyField,
    FieldMany2One: FieldMany2One,
};
```

as in the enterprise version. Therefore the module can't work in 9c.

**Comment by sbidoul** [comment]
@lmignon can you do the PR to Odoo? Can you also check with @adrienpeiffer I think he has done a similar PR recently.

**Comment by lmignon** [comment]
@sbidoul Yes I can.

**Comment by zakiuu** [comment]
@lmignon Also the M2ODialog, that's why I had to rewrite the whole code part related to it

**Comment by lmignon** [comment]
This PR depends of https://github.com/odoo/odoo/pull/13518

**Comment by pedrobaeza** [comment]
I had the same issue with one2many field and what I do is:

```
var FieldOne2Many = core.form_widget_registry.get('one2many');
```

File: web_m2x_options/static/src/js/form.js
Comment (njeudy) [reply web_m2x_options/static/src/js/form.js]: should use upstream one instead ?

**Comment by lmignon** [comment]
@pedrobaeza I've implemented your proposal to make the module compatible with Odoo 9.c without the merge of the fix proposed to odoo in odoo/odoo#13518

File: web_m2x_options/static/src/js/form.js
Comment (pedrobaeza) [inline web_m2x_options/static/src/js/form.js]: s/FieldOne2Many/FielMany2One (in two places)

**Comment by lmignon** [comment]
@pedrobaeza @njeudy changes done

**Comment by pedrobaeza** [comment]
Please squash commits by author and check Travis status.

**Comment by lmignon** [comment]
@pedrobaeza Errors in travis are in `web_widget_float_formula` I've already rebased the branch on 9.0...

**Comment by pedrobaeza** [comment]
OK, thanks, what about the squashing? Is that possible?

**Comment by lmignon** [comment]
@pedrobaeza done

**Comment by pedrobaeza** [comment]
@Yajo, can you help us about the fail in the JS test?

**Comment by yajo** [comment]
@lmignon is right; the guilty addon is `web_widget_float_formula`.

This is most probably one of the expected consequences of https://github.com/odoo/odoo/pull/13743: the test was probably failing before, but nobody noticed because Travis was getting fake :white_check_mark: status.

Another possibility is that this PR introduces the failure, and that's really possible because last builds from stable branches are all green... Could you check if your addon introduces any incompatibility, please @lmignon?

**Comment by pedrobaeza** [comment]
@Yajo, I don't think so, because it fails in other PRs too. I have restarted the build in CI to be sure if in main branch is also incorrect. Are you able to provide a fix to get green Travis again?

**Comment by yajo** [comment]
Right now I'm stuck in another task, I hope I can soon.

**Comment by lmignon** [comment]
@pedrobaeza @Yajo Therefore, this one can be  reviewed and merged..

**Comment by yajo** [comment]
#460 should fix your problem.

**Comment by yajo** [comment]
#460 merged, please rebase

**Comment by lmignon** [comment]
thank you @Yajo Rebase done...

**Comment by pedrobaeza** [comment]
Can you please include https://github.com/OCA/web/pull/434 and squash commits by author and logic?

**Comment by lmignon** [comment]
@pedrobaeza sorry but we spent too much time on this PR. I've already squashed the commit for members of Acsone and now you request the inclusion of a PR made after this one. Changes of #434  should be proposed in another PR and can not block this one.

**Comment by pedrobaeza** [comment]
OK, @lmignon, I'm not blocking it. I'm asking if you do the favor of including that interesting PR, but if you don't want, I can't force you. Anyway, I would ask again for a better squashing, as I still continue seeing a lot of commits from "OCA Transbot" or Holger.

**Comment by lmignon** [comment]
@pedrobaeza The commits from Holger and OCA Transbot are those from 8.0 (IMO we must keep these commits). The cherry-pick of #434 introduces conflicts and I've no time to try to resolve these conflicts (and to do the appropriate tests once merged to be sure that everything still works). 
My main concern is that PR after PR I  wonder if the balance benefits/costs remains positive when contributing to OCA....

**Comment by pedrobaeza** [comment]
@lmignon the idea after squashing commits is to keep a clean commit history in each branch, which is treated isolated from previous versions. Doing this, we reduce the git size, ease searches through history and so on. This is something not about OCA, but code version management in general. If you don't want to do it, I can do it, but the PR then will count as closed instead of merge (as I have to change commit history before pushing).

**Comment by lmignon** [comment]
@pedrobaeza I strongly disagree to squash commits from others branches. Keeping the same commits is the easiest way to compare 2 branches and detect which commits are missing. (except OCA transbot)

**Review by adrienpeiffer — APPROVED** [review]
:+1:

**Comment by pedrobaeza** [comment]
OK, I don't want to push too much. Thanks for the squash.

Merging...

**Comment by jlzhou** [comment]
#482 Migrate to odoo 10.0, add new 'open' option to control external button next to drop down.

**Pull request merged**
